require('./ProfileDataLogic.spec')
require('./AdminLogic.spec')
require('./ScoreLogic.spec')
require('./NumberFormat.spec')
require('./ZoneLogic.spec')
require('./LockedZonePenaltyLogic.spec')
require('./ZoneProgressionLogic.spec')
require('./ScorePopupLogic.spec')
require('./ZoneUnlockUiLogic.spec')
require('./WorldDecorLogic.spec')
require('./WorldLayoutExportLogic.spec')
require('./WorldLayoutSyncFlowLogic.spec')
require('./RankProgressionLogic.spec')
require('./LuckyChestLogic.spec')
require('./RewardCatalogLogic.spec')
require('./EquipmentLogic.spec')
require('./RankAuraLogic.spec')
print('All tests passed.')
